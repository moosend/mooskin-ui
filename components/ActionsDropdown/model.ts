import { IInputCallbackData } from '../_utils/types/commonTypes';
import { IBoxComponentProps } from '../Box/model';

export interface IActionsDropdownComponentProps extends IBoxComponentProps {
	/** whether the dropdown is open or not */
	isOpen?: boolean;

	/** callback called when clicking a ActionsDropdown item */
	onClickItem?: (e: React.MouseEvent<HTMLElement>, data: IInputCallbackData) => void;
}

export interface IActionsDropdownItemComponentProps extends IBoxComponentProps {
	/** data label returned in callback fn */
	dataLabel?: any;

	/** value returned in callback fn */
	value?: any;
}
