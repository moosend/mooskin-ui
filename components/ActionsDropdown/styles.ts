import styled, { keyframes } from 'styled-components';

// Models
import { IBoxComponentProps } from '../index';
import { IActionsDropdownComponentProps, IActionsDropdownItemComponentProps } from './model';

// Components
import { Box } from '../Box/Box';

// "CSS" variables
import variables from '../_utils/globals/variables';

const fadeIn = keyframes`
 0% {
	 opacity: 0;
	 transform: translate(0, -20px);
 }
 100% {
	 opacity: 1;
	 transform: translate(0);
 }
`;

export const StyledActionsDropdown = styled(Box)<IActionsDropdownComponentProps>`
	border-radius: 2px;
	background-color: ${(props) => props.palette?.backgroundColors.primary1 || variables.backgroundColors.primary1};
	display: flex;
	flex-direction: column;
	position: relative;
	z-index: 2;
`;

StyledActionsDropdown.displayName = 'StyledActionsDropdown';

export const StyledActionsDropdownFadeIn = styled(StyledActionsDropdown)`
	animation: ${fadeIn} 0.15s;
`;

StyledActionsDropdownFadeIn.displayName = 'StyledActionsDropdownFadeIn';

export const StyledActionsDropdownFadeOut = styled(StyledActionsDropdown)`
	opacity: 0;
	animation: ${fadeIn} 0.15s reverse;
`;

StyledActionsDropdownFadeOut.displayName = 'StyledActionsDropdownFadeOut';

export const StyledActionsDropdownItem = styled(Box)<IActionsDropdownItemComponentProps>`
	font-size: 12px;
	font-weight: bold;
	font-stretch: normal;
	font-style: normal;
	letter-spacing: normal;
	text-align: left;
	white-space: nowrap;
	color: ${(props) => props.palette?.fontColors.white || variables.fontColors.white};
	cursor: pointer;
	&:not(:last-child) {
		padding: 10px 10px 0;
	}
	&:last-child {
		padding: 10px;
	}
`;

StyledActionsDropdownItem.displayName = 'StyledActionsDropdownItem';

export const StyledActionsDropdownArrow = styled(Box)<IBoxComponentProps>`
	position: absolute;
	top: -6px;
	right: 22px;
	width: 12px;
	height: 12px;
	transform: rotate(45deg);
	z-index: -1;
	background-color: ${(props) => props.palette?.backgroundColors.primary1 || variables.backgroundColors.primary1};
`;

StyledActionsDropdownArrow.displayName = 'StyledActionsDropdownArrow';
