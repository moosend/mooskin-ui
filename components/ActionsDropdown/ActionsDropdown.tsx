import * as React from 'react';

// Mooskin Context HoC that passes context to component props
import { withMooskinContext } from '../Styled/MooskinContextProvider';

// Models
import { IInputCallbackData } from '../_utils/types/commonTypes';
import { Box } from '../Box/Box';
import { IBoxComponentProps } from '../Box/model';
import { IActionsDropdownComponentProps, IActionsDropdownItemComponentProps } from './model';

// Styled Components
import { StyledActionsDropdownArrow, StyledActionsDropdownFadeIn, StyledActionsDropdownFadeOut, StyledActionsDropdownItem } from './styles';

// Transitions
import { Transition } from 'react-transition-group';
const DropdownComponents = {
	entered: StyledActionsDropdownFadeIn,
	entering: StyledActionsDropdownFadeIn,
	exited: null,
	exiting: StyledActionsDropdownFadeOut,
	unmounted: null,
};

/**
 * ActionsDropdown
 */
export const ActionsDropdown: React.FC<IActionsDropdownComponentProps> = withMooskinContext((props) => {
	const [hasArrow, setHasArrow] = React.useState(false);

	const batchClickHandler = (
		e: React.MouseEvent<HTMLElement>,
		data: IInputCallbackData,
		callback?: (e: React.MouseEvent<HTMLElement>, data: IInputCallbackData) => void
	) => {
		props.onClickItem && props.onClickItem(e, data);
		callback && callback(e, data);
	};

	const recurseChildren = (children: any): any => {
		if (!children) {
			return null;
		}

		return React.Children.map(children, (child, i) => {
			if (React.isValidElement<IActionsDropdownItemComponentProps>(child) && child.type === ActionsDropdownItem) {
				return React.cloneElement(child, {
					children: recurseChildren((child.props as any).children),
					key: i,
					onClick: (e) => batchClickHandler(e, { dataLabel: child.props.dataLabel, value: child.props.value }, child.props.onClick),
				} as IActionsDropdownItemComponentProps);
			}

			if (React.isValidElement<IBoxComponentProps>(child) && child.type === ActionsDropdownArrow) {
				!hasArrow && setHasArrow(true);
				return React.cloneElement(child, {
					children: recurseChildren((child.props as any).children),
					key: i,
				} as IBoxComponentProps);
			}

			if (React.isValidElement(child) && (child.props as any).children) {
				return React.cloneElement(child, { key: i, children: recurseChildren((child.props as any).children) } as any);
			}

			return child;
		});
	};

	return (
		<Transition addEndListener={() => undefined} unmountOnExit in={props.isOpen} timeout={150}>
			{(state) => {
				const ActionDropdownComponent = DropdownComponents[state];
				if (ActionDropdownComponent) {
					return (
						<ActionDropdownComponent boxShadow="base" {...props}>
							<Box position="absolute" h={30} top={-30} left={0} right={0} />
							{!hasArrow && <ActionsDropdownArrow boxShadow="base" />}
							{recurseChildren(props.children)}
						</ActionDropdownComponent>
					);
				}
				return null;
			}}
		</Transition>
	);
});

ActionsDropdown.defaultProps = {
	className: '',
	style: {},
};

ActionsDropdown.displayName = 'ActionsDropdown';

/**
 * ActionsDropdownItem
 */
export const ActionsDropdownItem: React.FC<IActionsDropdownItemComponentProps> = withMooskinContext((props) => {
	return <StyledActionsDropdownItem {...props} />;
});

ActionsDropdownItem.defaultProps = {
	className: '',
	style: {},
};

ActionsDropdownItem.displayName = 'ActionsDropdownItem';

/**
 * ActionsDropdownArrow
 */
export const ActionsDropdownArrow: React.FC<IBoxComponentProps> = withMooskinContext((props) => {
	return <StyledActionsDropdownArrow {...props} />;
});

ActionsDropdownArrow.defaultProps = {
	className: '',
	style: {},
};

ActionsDropdownArrow.displayName = 'ActionsDropdownArrow';
