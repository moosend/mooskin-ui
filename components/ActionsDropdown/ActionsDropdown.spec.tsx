import * as React from 'react';

import { ActionsDropdown, ActionsDropdownItem } from './ActionsDropdown';

import { mount } from 'enzyme';

describe('ActionsDropdown', () => {
	test('renders ActionsDropdown correctly', () => {
		const func = jest.fn();

		const tree = mount(
			<ActionsDropdown onClickItem={func}>
				<ActionsDropdownItem dataLabel="settings" value="/settings">
					Settings
				</ActionsDropdownItem>
				<ActionsDropdownItem dataLabel="template" value="/template">
					Template
				</ActionsDropdownItem>
				<ActionsDropdownItem dataLabel="preview" value="/preview">
					Preview
				</ActionsDropdownItem>
			</ActionsDropdown>
		);
		expect(tree).toMatchSnapshot();
	});

	test('renders ActionsDropdownItem correctly', () => {
		const func = jest.fn();

		const tree = mount(<ActionsDropdownItem onClick={func} className="myClass" style={{ color: 'blue' }} />);
		expect(tree).toMatchSnapshot();
	});

	test.skip('calls click callbacks correctly', () => {
		const func = jest.fn();

		const tree = mount(
			<ActionsDropdown onClickItem={func}>
				<ActionsDropdownItem value="/settings">Settings</ActionsDropdownItem>
				<ActionsDropdownItem value="/template">Template</ActionsDropdownItem>
				<ActionsDropdownItem value="/preview">Preview</ActionsDropdownItem>
			</ActionsDropdown>
		);

		tree.find('div').at(2).simulate('click');
		expect(func).toHaveBeenCalled();
	});
});
