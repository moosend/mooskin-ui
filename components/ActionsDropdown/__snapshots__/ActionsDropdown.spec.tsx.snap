// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionsDropdown renders ActionsDropdown correctly 1`] = `
<ActionsDropdown
  className=""
  onClickItem={[MockFunction]}
  style={Object {}}
>
  <Component
    className=""
    onClickItem={[MockFunction]}
    palette={
      Object {
        "backgroundColors": Object {
          "background": "#FAFAFA",
          "black": "#000000",
          "darkgray1": "#777777",
          "darkgray2": "#454545",
          "gray1": "#E0E0E0",
          "gray2": "#E7E7E7",
          "medgray1": "#757575",
          "medgray2": "#888888",
          "primary1": "#4169E1",
          "primary2": "#212121",
          "primary3": "#00796B",
          "primary4": "#D32F2F",
          "secondary1": "#2E7D32",
          "secondary2": "#757575",
          "secondary3": "#673AB7",
          "secondary4": "#9C27B0",
          "secondary5": "#00796B",
          "secondary6": "#ECF0FC",
          "text": "#212121",
          "white": "#FFFFFF",
        },
        "borderColors": Object {
          "background": "#FAFAFA",
          "black": "#000000",
          "darkgray1": "#777777",
          "darkgray2": "#454545",
          "gray1": "#E0E0E0",
          "gray2": "#E7E7E7",
          "medgray1": "#757575",
          "medgray2": "#888888",
          "primary1": "#4169E1",
          "primary2": "#212121",
          "primary3": "#00796B",
          "primary4": "#D32F2F",
          "secondary1": "#2E7D32",
          "secondary2": "#757575",
          "secondary3": "#673AB7",
          "secondary4": "#9C27B0",
          "secondary5": "#00796B",
          "secondary6": "#ECF0FC",
          "text": "#212121",
          "white": "#FFFFFF",
        },
        "fontColors": Object {
          "background": "#FAFAFA",
          "black": "#000000",
          "darkgray1": "#777777",
          "darkgray2": "#454545",
          "gray1": "#E0E0E0",
          "gray2": "#E7E7E7",
          "medgray1": "#757575",
          "medgray2": "#888888",
          "primary1": "#4169E1",
          "primary2": "#212121",
          "primary3": "#00796B",
          "primary4": "#D32F2F",
          "secondary1": "#2E7D32",
          "secondary2": "#757575",
          "secondary3": "#673AB7",
          "secondary4": "#9C27B0",
          "secondary5": "#00796B",
          "secondary6": "#ECF0FC",
          "text": "#212121",
          "white": "#FFFFFF",
        },
      }
    }
    style={Object {}}
  />
</ActionsDropdown>
`;

exports[`ActionsDropdown renders ActionsDropdownItem correctly 1`] = `
<ActionsDropdownItem
  className="myClass"
  onClick={[MockFunction]}
  style={
    Object {
      "color": "blue",
    }
  }
>
  <Component
    className="myClass"
    onClick={[MockFunction]}
    palette={
      Object {
        "backgroundColors": Object {
          "background": "#FAFAFA",
          "black": "#000000",
          "darkgray1": "#777777",
          "darkgray2": "#454545",
          "gray1": "#E0E0E0",
          "gray2": "#E7E7E7",
          "medgray1": "#757575",
          "medgray2": "#888888",
          "primary1": "#4169E1",
          "primary2": "#212121",
          "primary3": "#00796B",
          "primary4": "#D32F2F",
          "secondary1": "#2E7D32",
          "secondary2": "#757575",
          "secondary3": "#673AB7",
          "secondary4": "#9C27B0",
          "secondary5": "#00796B",
          "secondary6": "#ECF0FC",
          "text": "#212121",
          "white": "#FFFFFF",
        },
        "borderColors": Object {
          "background": "#FAFAFA",
          "black": "#000000",
          "darkgray1": "#777777",
          "darkgray2": "#454545",
          "gray1": "#E0E0E0",
          "gray2": "#E7E7E7",
          "medgray1": "#757575",
          "medgray2": "#888888",
          "primary1": "#4169E1",
          "primary2": "#212121",
          "primary3": "#00796B",
          "primary4": "#D32F2F",
          "secondary1": "#2E7D32",
          "secondary2": "#757575",
          "secondary3": "#673AB7",
          "secondary4": "#9C27B0",
          "secondary5": "#00796B",
          "secondary6": "#ECF0FC",
          "text": "#212121",
          "white": "#FFFFFF",
        },
        "fontColors": Object {
          "background": "#FAFAFA",
          "black": "#000000",
          "darkgray1": "#777777",
          "darkgray2": "#454545",
          "gray1": "#E0E0E0",
          "gray2": "#E7E7E7",
          "medgray1": "#757575",
          "medgray2": "#888888",
          "primary1": "#4169E1",
          "primary2": "#212121",
          "primary3": "#00796B",
          "primary4": "#D32F2F",
          "secondary1": "#2E7D32",
          "secondary2": "#757575",
          "secondary3": "#673AB7",
          "secondary4": "#9C27B0",
          "secondary5": "#00796B",
          "secondary6": "#ECF0FC",
          "text": "#212121",
          "white": "#FFFFFF",
        },
      }
    }
    style={
      Object {
        "color": "blue",
      }
    }
  >
    <StyledActionsDropdownItem
      className="myClass"
      onClick={[MockFunction]}
      palette={
        Object {
          "backgroundColors": Object {
            "background": "#FAFAFA",
            "black": "#000000",
            "darkgray1": "#777777",
            "darkgray2": "#454545",
            "gray1": "#E0E0E0",
            "gray2": "#E7E7E7",
            "medgray1": "#757575",
            "medgray2": "#888888",
            "primary1": "#4169E1",
            "primary2": "#212121",
            "primary3": "#00796B",
            "primary4": "#D32F2F",
            "secondary1": "#2E7D32",
            "secondary2": "#757575",
            "secondary3": "#673AB7",
            "secondary4": "#9C27B0",
            "secondary5": "#00796B",
            "secondary6": "#ECF0FC",
            "text": "#212121",
            "white": "#FFFFFF",
          },
          "borderColors": Object {
            "background": "#FAFAFA",
            "black": "#000000",
            "darkgray1": "#777777",
            "darkgray2": "#454545",
            "gray1": "#E0E0E0",
            "gray2": "#E7E7E7",
            "medgray1": "#757575",
            "medgray2": "#888888",
            "primary1": "#4169E1",
            "primary2": "#212121",
            "primary3": "#00796B",
            "primary4": "#D32F2F",
            "secondary1": "#2E7D32",
            "secondary2": "#757575",
            "secondary3": "#673AB7",
            "secondary4": "#9C27B0",
            "secondary5": "#00796B",
            "secondary6": "#ECF0FC",
            "text": "#212121",
            "white": "#FFFFFF",
          },
          "fontColors": Object {
            "background": "#FAFAFA",
            "black": "#000000",
            "darkgray1": "#777777",
            "darkgray2": "#454545",
            "gray1": "#E0E0E0",
            "gray2": "#E7E7E7",
            "medgray1": "#757575",
            "medgray2": "#888888",
            "primary1": "#4169E1",
            "primary2": "#212121",
            "primary3": "#00796B",
            "primary4": "#D32F2F",
            "secondary1": "#2E7D32",
            "secondary2": "#757575",
            "secondary3": "#673AB7",
            "secondary4": "#9C27B0",
            "secondary5": "#00796B",
            "secondary6": "#ECF0FC",
            "text": "#212121",
            "white": "#FFFFFF",
          },
        }
      }
      style={
        Object {
          "color": "blue",
        }
      }
    >
      <Box
        className="sc-eCApGN kpqzvj myClass"
        onClick={[MockFunction]}
        palette={
          Object {
            "backgroundColors": Object {
              "background": "#FAFAFA",
              "black": "#000000",
              "darkgray1": "#777777",
              "darkgray2": "#454545",
              "gray1": "#E0E0E0",
              "gray2": "#E7E7E7",
              "medgray1": "#757575",
              "medgray2": "#888888",
              "primary1": "#4169E1",
              "primary2": "#212121",
              "primary3": "#00796B",
              "primary4": "#D32F2F",
              "secondary1": "#2E7D32",
              "secondary2": "#757575",
              "secondary3": "#673AB7",
              "secondary4": "#9C27B0",
              "secondary5": "#00796B",
              "secondary6": "#ECF0FC",
              "text": "#212121",
              "white": "#FFFFFF",
            },
            "borderColors": Object {
              "background": "#FAFAFA",
              "black": "#000000",
              "darkgray1": "#777777",
              "darkgray2": "#454545",
              "gray1": "#E0E0E0",
              "gray2": "#E7E7E7",
              "medgray1": "#757575",
              "medgray2": "#888888",
              "primary1": "#4169E1",
              "primary2": "#212121",
              "primary3": "#00796B",
              "primary4": "#D32F2F",
              "secondary1": "#2E7D32",
              "secondary2": "#757575",
              "secondary3": "#673AB7",
              "secondary4": "#9C27B0",
              "secondary5": "#00796B",
              "secondary6": "#ECF0FC",
              "text": "#212121",
              "white": "#FFFFFF",
            },
            "fontColors": Object {
              "background": "#FAFAFA",
              "black": "#000000",
              "darkgray1": "#777777",
              "darkgray2": "#454545",
              "gray1": "#E0E0E0",
              "gray2": "#E7E7E7",
              "medgray1": "#757575",
              "medgray2": "#888888",
              "primary1": "#4169E1",
              "primary2": "#212121",
              "primary3": "#00796B",
              "primary4": "#D32F2F",
              "secondary1": "#2E7D32",
              "secondary2": "#757575",
              "secondary3": "#673AB7",
              "secondary4": "#9C27B0",
              "secondary5": "#00796B",
              "secondary6": "#ECF0FC",
              "text": "#212121",
              "white": "#FFFFFF",
            },
          }
        }
        style={
          Object {
            "color": "blue",
          }
        }
        transition="0.2s all"
      >
        <Component
          className="sc-eCApGN kpqzvj myClass"
          onClick={[MockFunction]}
          palette={
            Object {
              "backgroundColors": Object {
                "background": "#FAFAFA",
                "black": "#000000",
                "darkgray1": "#777777",
                "darkgray2": "#454545",
                "gray1": "#E0E0E0",
                "gray2": "#E7E7E7",
                "medgray1": "#757575",
                "medgray2": "#888888",
                "primary1": "#4169E1",
                "primary2": "#212121",
                "primary3": "#00796B",
                "primary4": "#D32F2F",
                "secondary1": "#2E7D32",
                "secondary2": "#757575",
                "secondary3": "#673AB7",
                "secondary4": "#9C27B0",
                "secondary5": "#00796B",
                "secondary6": "#ECF0FC",
                "text": "#212121",
                "white": "#FFFFFF",
              },
              "borderColors": Object {
                "background": "#FAFAFA",
                "black": "#000000",
                "darkgray1": "#777777",
                "darkgray2": "#454545",
                "gray1": "#E0E0E0",
                "gray2": "#E7E7E7",
                "medgray1": "#757575",
                "medgray2": "#888888",
                "primary1": "#4169E1",
                "primary2": "#212121",
                "primary3": "#00796B",
                "primary4": "#D32F2F",
                "secondary1": "#2E7D32",
                "secondary2": "#757575",
                "secondary3": "#673AB7",
                "secondary4": "#9C27B0",
                "secondary5": "#00796B",
                "secondary6": "#ECF0FC",
                "text": "#212121",
                "white": "#FFFFFF",
              },
              "fontColors": Object {
                "background": "#FAFAFA",
                "black": "#000000",
                "darkgray1": "#777777",
                "darkgray2": "#454545",
                "gray1": "#E0E0E0",
                "gray2": "#E7E7E7",
                "medgray1": "#757575",
                "medgray2": "#888888",
                "primary1": "#4169E1",
                "primary2": "#212121",
                "primary3": "#00796B",
                "primary4": "#D32F2F",
                "secondary1": "#2E7D32",
                "secondary2": "#757575",
                "secondary3": "#673AB7",
                "secondary4": "#9C27B0",
                "secondary5": "#00796B",
                "secondary6": "#ECF0FC",
                "text": "#212121",
                "white": "#FFFFFF",
              },
            }
          }
          style={
            Object {
              "color": "blue",
            }
          }
          transition="0.2s all"
        >
          <StyledBox
            className="sc-eCApGN kpqzvj myClass"
            onClick={[MockFunction]}
            palette={
              Object {
                "backgroundColors": Object {
                  "background": "#FAFAFA",
                  "black": "#000000",
                  "darkgray1": "#777777",
                  "darkgray2": "#454545",
                  "gray1": "#E0E0E0",
                  "gray2": "#E7E7E7",
                  "medgray1": "#757575",
                  "medgray2": "#888888",
                  "primary1": "#4169E1",
                  "primary2": "#212121",
                  "primary3": "#00796B",
                  "primary4": "#D32F2F",
                  "secondary1": "#2E7D32",
                  "secondary2": "#757575",
                  "secondary3": "#673AB7",
                  "secondary4": "#9C27B0",
                  "secondary5": "#00796B",
                  "secondary6": "#ECF0FC",
                  "text": "#212121",
                  "white": "#FFFFFF",
                },
                "borderColors": Object {
                  "background": "#FAFAFA",
                  "black": "#000000",
                  "darkgray1": "#777777",
                  "darkgray2": "#454545",
                  "gray1": "#E0E0E0",
                  "gray2": "#E7E7E7",
                  "medgray1": "#757575",
                  "medgray2": "#888888",
                  "primary1": "#4169E1",
                  "primary2": "#212121",
                  "primary3": "#00796B",
                  "primary4": "#D32F2F",
                  "secondary1": "#2E7D32",
                  "secondary2": "#757575",
                  "secondary3": "#673AB7",
                  "secondary4": "#9C27B0",
                  "secondary5": "#00796B",
                  "secondary6": "#ECF0FC",
                  "text": "#212121",
                  "white": "#FFFFFF",
                },
                "fontColors": Object {
                  "background": "#FAFAFA",
                  "black": "#000000",
                  "darkgray1": "#777777",
                  "darkgray2": "#454545",
                  "gray1": "#E0E0E0",
                  "gray2": "#E7E7E7",
                  "medgray1": "#757575",
                  "medgray2": "#888888",
                  "primary1": "#4169E1",
                  "primary2": "#212121",
                  "primary3": "#00796B",
                  "primary4": "#D32F2F",
                  "secondary1": "#2E7D32",
                  "secondary2": "#757575",
                  "secondary3": "#673AB7",
                  "secondary4": "#9C27B0",
                  "secondary5": "#00796B",
                  "secondary6": "#ECF0FC",
                  "text": "#212121",
                  "white": "#FFFFFF",
                },
              }
            }
            style={
              Object {
                "color": "blue",
              }
            }
            transition="0.2s all"
          >
            <div
              className="sc-bdnylx jknATe sc-eCApGN kpqzvj myClass"
              onClick={[MockFunction]}
              style={
                Object {
                  "color": "blue",
                }
              }
            />
          </StyledBox>
        </Component>
      </Box>
    </StyledActionsDropdownItem>
  </Component>
</ActionsDropdownItem>
`;
